src/celeritas/optical/ScintillationParams.cc

```diff
@@ -37,80 +40,225 @@ ScintillationParams::from_import(ImportData const& data)
-                           << " for scintillation component " << i
-                           << " (should be positive)");
+            auto const pdg = particle_params->id_to_pdg(ParticleId(pid)).get();
+            for (auto matid : range(num_mat))
+            {
+                // Particle- and material-dependend scintillation spectrum
+                auto iter = input.data[matid].particles.find(pdg);
+                CELER_ASSERT(iter != input.data[matid].particles.end()
+                             && pdg == iter->first);
+
+                auto const& ipss = iter->second;
+                ParticleScintillationSpectrum part_spec;
+                part_spec.yield_vector.grid = build_grid.insert_back(
+                    ipss.yield_vector.x.begin(), ipss.yield_vector.x.end());
+                part_spec.yield_vector.value = build_grid.insert_back(
+                    ipss.yield_vector.y.begin(), ipss.yield_vector.y.end());
+                CELER_ASSERT(part_spec.yield_vector);
+
+                auto comps = this->copy_components(ipss.components);
+                part_spec.components
+                    = build_compoments.insert_back(comps.begin(), comps.end());
+
+                build_particles.push_back(std::move(part_spec));
+            }
         }
-        ScintillationSpectrum spectrum;
-        spectrum.yield = spec.material.yield;
-        CELER_VALIDATE(spectrum.yield > 0,
-                       << "invalid yield=" << spectrum.yield
-                       << " for scintillation (should be positive)");
-        spectrum.resolution_scale = spec.resolution_scale;
-        CELER_VALIDATE(
-            spectrum.resolution_scale >= 0,
-            << "invalid resolution_scale=" << spectrum.resolution_scale
-            << " for scintillation (should be nonnegative)");
-        spectrum.components = components.insert_back(comp.begin(), comp.end());
-        spectra.push_back(spectrum);
+        CELER_ENSURE(build_particles.size() == num_part * num_mat);
     }
 
     mirror_ = CollectionMirror<ScintillationData>{std::move(host_data)};
     CELER_ENSURE(mirror_);
 }
 
+//---------------------------------------------------------------------------//
+/*
+ * Return a Celeritas ScintillationComponent from an imported one.
+ */
+std::vector<ScintillationComponent> ScintillationParams::copy_components(
+    std::vector<ImportScintComponent> const& input_comp)
+{
+    std::vector<ScintillationComponent> comp(input_comp.size());
+    real_type norm{0};
+    for (auto i : range(comp.size()))
+    {
+        comp[i].lambda_mean = input_comp[i].lambda_mean;
+        comp[i].lambda_sigma = input_comp[i].lambda_sigma;
+        comp[i].rise_time = input_comp[i].rise_time;
+        comp[i].fall_time = input_comp[i].fall_time;
+        norm += input_comp[i].yield;
+    }
+
+    // Store normalized yield
+    for (auto i : range(comp.size()))
+    {
+        comp[i].yield_frac = input_comp[i].yield / norm;
+    }
+
+    this->validate(comp);
```

Comment:
Just readability. I can keep everything inside `build_components()` if you think it's better.

Reply:
I think it'd be better to just move the validations inside the final loop. Defining a separate class member function that takes a vector makes the purpose less clear. If you want to move the validates to a separate place, put them in an anonymous namespace and take a single component rather than a vector, since nothing about the validation requires information about multiple components simultaneously.